pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -330,12 +371,42 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
 	}
 
 	reconciler.EmitEvent(c.Recorder, before, after, tr)
-
 	c.Logger.Infof("Successfully reconciled taskrun %s/%s with status: %#v", tr.Name, tr.Namespace, after)
 
 	return nil
 }
 
+func (c *Reconciler) sendCloudEvents(tr *v1alpha1.TaskRun) error {
+	// The TaskRun is complete. Time to send cloud events (if any).
+	sendErrors := 0
+	var err error
+	for idx, cloudEventDelivery := range tr.Status.CloudEvents {
+		eventStatus := &(tr.Status.CloudEvents[idx].Status)
+		// Skip events that have already been sent (successfully or unsuccessfully)
+		// Ensure we try to send all events once (possibly through different reconcile calls)
+		if eventStatus.Condition != v1alpha1.CloudEventConditionUnknown || eventStatus.RetryCount > 0 {
+			continue
+		}
+		_, err := resources.SendTaskRunCloudEvent(cloudEventDelivery.Target, tr, c.Logger, c.cloudEventClientFactory)
+		eventStatus.SentAt = &metav1.Time{Time: time.Now()}
+		eventStatus.RetryCount = eventStatus.RetryCount + 1
+		if err != nil {
+			eventStatus.Condition = v1alpha1.CloudEventConditionFailed
+			eventStatus.Error = err.Error()
+			sendErrors++
+		} else {
+			c.Logger.Infof("Sent event for target %s", cloudEventDelivery.Target)
+			eventStatus.Condition = v1alpha1.CloudEventConditionSent
+		}
+	}
+	if sendErrors > 0 {
+		c.Logger.Errorf("Failed to send %d cloud events for TaskRun %s", sendErrors, tr.Name)
+		// Return the latest send error
+		return err
```

Comment:
This returns only the error from the last error delivery.  Isn't it better to return the multierror.Error?

Reply:
Nice point, I only discovered multierror after I wrote this code - so I will adapt this one now to send all errors back.